pkg/components/prometheus-operator/component_conversion_test.go

```diff
@@ -0,0 +1,171 @@
+package prometheus //nolint:testpackage
+
+import (
+	"reflect"
+	"testing"
+
+	"github.com/hashicorp/hcl/v2"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	yamlserializer "k8s.io/apimachinery/pkg/runtime/serializer/yaml"
+	"k8s.io/client-go/util/jsonpath"
+
+	"github.com/kinvolk/lokomotive/pkg/components/util"
+)
+
+func renderManifests(t *testing.T, configHCL string) map[string]string {
+	name := "prometheus-operator"
+
+	component := newComponent()
+
+	body, diagnostics := util.GetComponentBody(configHCL, name)
+	if diagnostics != nil {
+		t.Fatalf("Getting component body: %v", diagnostics)
+	}
+
+	diagnostics = component.LoadConfig(body, &hcl.EvalContext{})
+	if diagnostics.HasErrors() {
+		t.Fatalf("Valid config should not return an error, got: %v", diagnostics)
+	}
+
+	ret, err := component.RenderManifests()
+	if err != nil {
+		t.Fatalf("Rendering manifests with valid config should succeed, got: %s", err)
+	}
+
+	return ret
+}
+
+//nolint:funlen
+func TestConversion(t *testing.T) {
+	testCases := []struct {
+		Name                   string
+		InputConfig            string
+		ExpectedConfigFileName string
+		Expected               reflect.Value
+		JSONPath               string
+	}{
```

Comment:
Hmm, perhaps we could put that into test utils in the first place, so it can be also used for testing for other packages.

Reply:
Yes we can do that, once folks agree this is the right way to go forward. WDYT @johananl .

Reply:
You folks know my approach regarding "utils" :-)

@invidian what do you mean by "that" and "it"?

Reply:
I meant struct and a function consuming it. While I agree that for lokomotive related code we shouldn't use util packages, as they do not provide any context or information, I think for test code it's completely fine to have shared helpers, which makes writing tests easier and it would be intentional to not build any abstraction about it.

Reply:
Are you talking about making this test a generic test which can be used to test arbitrary components? If there is a *function* which can save lots of duplication when reused, it could make sense to make it a generic helper. I wouldn't put test cases in a central place though as they are specific to the "thing under test", in this case a component.

Reply:
>Are you talking about making this test a generic test which can be used to test arbitrary components?

Yes.

Reply:
In general, I'd rather keep things as simple as possible. I'll usually prefer simplicity over almost anything. This means that I'll be in favor for making some function generic and reuse it only when doing so **actually simplifies** the code.

It's a trade-off: having to jump around source files to figure out what some test is doing is more complex than having everything "right there" in front of you in a single file. On the other hand having to edit 100 test files because we've changed something could hint that we should be reusing something.

Reply:
> > Are you talking about making this test a generic test which can be used to test arbitrary components?
> 
> Yes.

Honestly I don't see a strong reason to do this right now. I would start thinking about making things more generic once I see we have an actual opportunity to reuse. I wouldn't make things more complex before then.

Reply:
I think we could easily re-use it in recently added Rook and MetalLB tests and possibly in all tests added in the future, if we start verifying rendered manifests. I think if we don't do it now, next tests will either follow content-based testing, which is more difficult to maintain or will require person implementing tests to make this code generic.

Reply:
Well, I was asked for my opinion. You have it now :slightly_smiling_face: :walking:

Reply:
@johananl And what do you think about using `unstructured`, `reflect` and JSON path in this tests? I think this is what @surajssd was referring to :smile:

Reply:
Oh, I interpreted @surajssd 's question as "should we do what you (@invidian) suggested?".

I admit I'm not sure what this test does. It's called `TestConversion` which doesn't give me much info and the implementation seems complex so I'd have to spend more time to figure out what it does.

Is it the same as [this](https://github.com/kinvolk/lokomotive/commit/39e3a898fa955e8493d8ffca187ca5736d6d1050) "conversion test"? If so, this link contains my preferred approach to this sort of tests. Tests cases contain input and output, nearly everything is done in the body of the test function, minimal use of fancy helper functions and jumping around (and even if we do - jump around only within the same source file).

Reply:
Let's put this in a shared place when we have another test that needs those functions and keep it here for now.

Regarding the JSON path support I like that the test cases are much shorter and I don't mind using a bit of magic in tests to get that.